### PR TITLE
[SPARK-34739][SQL] Support add/subtract of a year-month interval to/from a timestamp

### PR DESCRIPTION
**Pull request by MaxGekk**
### What changes were proposed in this pull request?
Support `timestamp +/- year-month interval`. In the PR, I propose to introduce new binary expression `TimestampAddYMInterval` similarly to `DateAddYMInterval`. It invokes new method `timestampAddMonths` from `DateTimeUtils` by passing a timestamp as an offset in microseconds since the epoch, amount of months from the giveb year-month interval, and the time zone ID in which the operation is performed. The `timestampAddMonths()` method converts the input microseconds to a local timestamp, adds months to it, and converts the results back to an instant in microseconds at the given time zone.

### Why are the changes needed?
To conform the ANSI SQL standard which requires to support such operation over timestamps and intervals:
Screenshot 2021-03-12 at 11 36 14

### Does this PR introduce _any_ user-facing change?
Should not since new intervals have not been released yet.

### How was this patch tested?
By running new tests:
```
$ build/sbt "test:testOnly *DateTimeUtilsSuite"
$ build/sbt "test:testOnly *DateExpressionsSuite"
$ build/sbt "test:testOnly *ColumnExpressionSuite"
```

### CODE REVIEW

**Comment by MaxGekk** [comment]
@cloud-fan @HyukjinKwon @yaooqinn Could you review this, please.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala [hunk 1]
```diff
@@ -1521,6 +1521,40 @@ case class DateAddYMInterval(date: Expression, interval: Expression) extends Add
   override def sql: String = s"${left.sql} + ${right.sql}"
 }
 
+// Adds the year-month interval to the timestamp
+case class TimestampAddYMInterval(
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1525]: Shall we make it general and accept both year-month and day-time intervals? The implementation can have 2 branches:
```
override def nullSafeEval(micros: Any, months: Any): Any = interval.dataType match {
  case YearMonthIntervalType =>
    timestampAddMonths(micros.asInstanceOf[Long], months.asInstanceOf[Int], zoneId)
  case DayTimeIntervalType => ...
}
```

Comment (MaxGekk) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/datetimeExpressions.scala:1525]: I plan to support `timestamp + day-time interval` separately in another PR. The implementation itself is not complex, need to write more tests.

**Review by cloud-fan — APPROVED** [review]

**Comment by MaxGekk** [comment]
Thanks @cloud-fan. I am merging this to master.

**Review by yaooqinn — APPROVED** [review]

**Pull request closed without merging**